### PR TITLE
azure-cli-2 0.1.6 (new formula)

### PR DESCRIPTION
**Pull request by derekbekoe**
Formula for Azure CLI 2.0 available at https://github.com/azure/azure-cli.

- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----
Closes https://github.com/Azure/azure-cli/issues/643
FYI @tjprescott, @yugangw-msft, @johanste, @troydai

### CODE REVIEW

File: Formula/azure-cli-2.rb [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+class AzureCli2 < Formula
+  include Language::Python::Virtualenv
+
+  desc "Azure CLI 2.0"
+  homepage "https://github.com/Azure/azure-cli"
+  url "https://azurecliprod.blob.core.windows.net/releases/azure-cli_packaged_0.1.5.tar.gz"
+  sha256 "cb65651530544c43343e2c80b0d96a681ad0a07367e288ef9fae261093c497f3"
+  depends_on "python"
+
+  # Apply the 'az component' patch
```
Comment (dunn) [inline Formula/azure-cli-2.rb]: Are these patches going to be included in a future release of Azure?

File: Formula/azure-cli-2.rb [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    sha256 "d61ef29ace9bbdfef9a25dfbb1f475225bbca174263c8f863ee70f87d0a78bbe"
+  end
+
+  # Apply pip module search patch
+  patch do
+    url "https://azurecliprod.blob.core.windows.net/patches/patch_0.1.5_pkg_util.diff"
+    sha256 "ea0879280dbb3074d464752c27bfe01a7673da14137a4c4315d1938a0d05a03e"
+  end
+
+  def completion_script; <<-EOS.undent
```
Comment (dunn) [inline Formula/azure-cli-2.rb]: Why isn't this distributed with the upstream tarball?

File: Formula/azure-cli-2.rb [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    bin.install_symlink "#{libexec}/bin/az"
+    # Bash tab completion
+    File.write(libexec/"az.completion", completion_script)
+    bash_completion.install libexec/"az.completion"
+  end
+
+  def caveats; <<-EOS.undent
+    To complete tab completion set up:
+      1. Modify your ~/.bash_profile (if the line doesn't already exist)
+        $ echo "[ -f #{HOMEBREW_PREFIX}/etc/bash_completion.d/az.completion ] && . #{HOMEBREW_PREFIX}/etc/bash_completion.d/az.completion" >> ~/.bash_profile
```
Comment (dunn) [inline Formula/azure-cli-2.rb]: This isn't necessary; when completion scripts are installed we automatically add a note to the caveats.

File: Formula/azure-cli-2.rb [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+      2. Restart your shell
+        $ exec -l $SHELL
+    ----
+    Get started with:
+      $ az configure
+    EOS
+  end
```
Comment (dunn) [inline Formula/azure-cli-2.rb]: None of the caveats are Homebrew-specific; please remove the entire block.

Comment (derekbekoe) [inline Formula/azure-cli-2.rb]: removed as suggested.

File: Formula/azure-cli-2.rb (hunk 3 above)
Comment (derekbekoe) [inline Formula/azure-cli-2.rb]: removed as suggested.

File: Formula/azure-cli-2.rb (hunk 1 above)
Comment (derekbekoe) [inline Formula/azure-cli-2.rb]: Upstream has the ability to update itself through commands such as `az component update`.
According to Homebrew documentation, ["We don’t like tools that upgrade themselves"](https://github.com/Homebrew/brew/blob/master/docs/Acceptable-Formulae.md#we-dont-like-tools-that-upgrade-themselves).
These two patches disable the self-updating functionality whilst still enabling upstream to support it as is desired.

File: Formula/azure-cli-2.rb (hunk 2 above)
Comment (derekbekoe) [inline Formula/azure-cli-2.rb]: Thanks for the suggestion.
Tracked in upstream with https://github.com/Azure/azure-cli/issues/1768.
This will be changed in the next release then this can be removed.

File: Formula/azure-cli-2.rb [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+class AzureCli2 < Formula
+  include Language::Python::Virtualenv
+
+  desc "Azure CLI 2.0"
+  homepage "https://github.com/Azure/azure-cli"
+  url "https://azurecliprod.blob.core.windows.net/releases/azure-cli_packaged_0.1.6.tar.gz"
+  sha256 "bb4bd4694e044bf416941e3afdc505bde7d049bac2e66c8b3fe44c9cb3b57ff8"
+  depends_on "python"
+
+  # Apply the 'az component' patch
```
Comment (MikeMcQuaid) [inline Formula/azure-cli-2.rb]: Has this patch been merged/submitted upstream? Regardless, can you link to the upstream merge/submission as a Ruby comment here, thanks!

File: Formula/azure-cli-2.rb [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  patch do
+    url "https://azurecliprod.blob.core.windows.net/patches/patch_2_component_custom.diff"
+    sha256 "d61ef29ace9bbdfef9a25dfbb1f475225bbca174263c8f863ee70f87d0a78bbe"
+  end
+
+  # Apply pip module search patch
```
Comment (MikeMcQuaid) [inline Formula/azure-cli-2.rb]: Has this patch been merged/submitted upstream? Regardless, can you link to the upstream merge/submission as a Ruby comment here, thanks!

File: Formula/azure-cli-2.rb [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  patch do
+    url "https://azurecliprod.blob.core.windows.net/patches/patch_2_pkg_util.diff"
+    sha256 "4b97507cb73b405c6fb2d701eb52ffa72ce547f791097fccadffc491ad6ae194"
+  end
+
+  def completion_script; <<-EOS.undent
```
Comment (MikeMcQuaid) [inline Formula/azure-cli-2.rb]: Can this be put somewhere upstream that we can download it instead?

File: Formula/azure-cli-2.rb [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+            unset COMPREPLY
+        fi
+    }
+    complete -o nospace -F _python_argcomplete "az"
+    EOS
+  end
+
+  def install
+    virtualenv_create(libexec)
+    Dir.mktmpdir do |tmp_dir|
```
Comment (MikeMcQuaid) [inline Formula/azure-cli-2.rb]: `install` is already in a `tmpdir`; why is this needed?

File: Formula/azure-cli-2.rb [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+      pkg_dirs = ["src/azure-cli", "src/azure-cli-core", "src/azure-cli-nspkg"]
+      pkg_dirs += Dir.glob("src/command_modules/azure-cli-*/")
```
Comment (MikeMcQuaid) [inline Formula/azure-cli-2.rb]: Use `Pathname.glob`, create `Pathname` objects above using `(buildpath/"src/azure-cli")` and then `path.cd` below.

File: Formula/azure-cli-2.rb [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+      # Build the packages
```
Comment (MikeMcQuaid) [inline Formula/azure-cli-2.rb]: This comment is self-explanatory so can be removed.

File: Formula/azure-cli-2.rb [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+      pkg_dirs.each do |item|
+        Dir.chdir(item) { system libexec/"bin/python", "setup.py", "bdist_wheel", "-d", tmp_dir }
+      end
+      # Install the CLI (with required modules only)
+      system libexec/"bin/pip", "install", "azure-cli", "-f", tmp_dir
+    end
+    # Write the executable
+    bin_dir = libexec/"bin"
+    File.write("#{bin_dir}/az", "#!/usr/bin/env bash\n#{bin_dir}/python -m azure.cli \"$@\"")
+    bin.install_symlink "#{libexec}/bin/az"
```
Comment (MikeMcQuaid) [inline Formula/azure-cli-2.rb]: May as well write this to `bin` directly with `(bin/"az").write` and use `EOS.undent` rather than `\n`

File: Formula/azure-cli-2.rb [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    # Bash tab completion
+    File.write(libexec/"az.completion", completion_script)
```
Comment (MikeMcQuaid) [inline Formula/azure-cli-2.rb]: Write this directly to `bash_completion/"az.completion"`.

File: Formula/azure-cli-2.rb [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    bash_completion.install libexec/"az.completion"
+  end
+
+  test do
+    bin_dir = libexec/"bin"
+    system bin_dir/"az", "--version"
```
Comment (MikeMcQuaid) [inline Formula/azure-cli-2.rb]: `system bin/"az", "--version` instead. Also, ideally have this do something rather than just query the version.

File: Formula/azure-cli-2.rb [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+    assert_match "complete -o nospace -F _python_argcomplete az",
```
Comment (MikeMcQuaid) [inline Formula/azure-cli-2.rb]: This seems pretty brittle and implementation dependent.

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]

File: Formula/azure-cli-2.rb (hunk 8 above)
Comment (derekbekoe) [inline Formula/azure-cli-2.rb]: Aim was to store the build files in a separate tmp dir but using `buildpath` is sufficient. Thanks.

File: Formula/azure-cli-2.rb (hunk 10 above)
Comment (derekbekoe) [inline Formula/azure-cli-2.rb]: removed.

File: Formula/azure-cli-2.rb (hunk 11 above)
Comment (derekbekoe) [inline Formula/azure-cli-2.rb]: done.

File: Formula/azure-cli-2.rb (hunk 12 above)
Comment (derekbekoe) [inline Formula/azure-cli-2.rb]: done.

File: Formula/azure-cli-2.rb (hunk 13 above)
Comment (derekbekoe) [inline Formula/azure-cli-2.rb]: done.

File: Formula/azure-cli-2.rb (hunk 14 above)
Comment (derekbekoe) [inline Formula/azure-cli-2.rb]: removed.

**Comment by derekbekoe** [comment]
Working with upstream to have the patches and completion script included there.
Will close this PR for now.
Thanks for the reviews and comments!

**Comment by balupton** [comment]
What's the latest on this? Seems no new progress was made:

```
> brew install azure-cli-2
Updating Homebrew...
Error: No available formula with the name "azure-cli-2" 
==> Searching for a previously deleted formula...
Warning: homebrew/core is shallow clone. To get complete history run:
  git -C "$(brew --repo homebrew/core)" fetch --unshallow

Error: No previously deleted formula found.
==> Searching for similarly named formulae...
Error: No similarly named formulae found.
==> Searching taps...
Error: No formulae found in taps.
```

**Comment by balupton** [comment]
For those wondering, this PR got closed in favour of https://github.com/Homebrew/homebrew-core/pull/13198
